### ISSUE TITLE
Races between multiple worker threads - random API failures

### ISSUE THREAD

**Issue by wipawel** [Vector35/binaryninja-api#2526]
Binary Ninja Version: 2.4.2875-dev, adb7c635
Platform: macOS 10.15

When running a standalone script (commercial license) using binaryninja API and multiple worker threads, the API calls start to behave non-deterministically. Sometimes, certain objects like func.low_level_il lose reference to its medium_level_il or vice versa. Other times, the func.medium_level_il.ssa_form[x] failes for index x, while it should not.

The problematic situation is transient and does not occur every time, but the script fails when it does.

Initially I though there is a race condition between multiple worker threads, where one thread performs analysis update and the other starts using the objects. So, I added bv.set_analysis_hold(True) to stop that. It did not help.

Then, I set the number of worker threads to 1. The issue is not reproducible since then.

**Comment by rssor** [comment]
Are you holding a reference to the initial IL function object once you start working with it, or are you grabbing it from the top level `Function` object every time?

Essentially, the IL function objects are handled as a snapshot. Once you grab and hold onto a reference to any of them, the related IL functions will persist and you can interact with them/query them safely. So going from say, `low_level_il_func.ssa_form[x]` or `low_level_il_func[x].ssa_form` should be safe, it relies on you having a persistent reference to one of these IL function objects. Similarly, jumping between levels from IL function objects should be safe: `low_level_il_func.medium_level_il` for example.

Native function objects are different, and when you query for the IL function object from one of them you'll be given the IL function object that was generated most recently. So if you have some long-running analysis and you are accessing `native_func.low_level_il[x]` repeatedly, all of those are opportunities for you to be getting results from a newer collection of IL function objects. Something that might result in you seeing results from multiple analysis results at once might be `low_level_il_func.source_function.medium_level_il`, because the original `low_level_il` might be much older than the new MLIL function object you get.

If you were to restructure your code slightly to hold onto one of the IL function objects and only get access to the other IL forms from that first IL function object I think your issues would go away.

In summary: try to only grab one of the IL functions from the native Function object once. Then get to the other IL forms by jumping directly between the IL functions, not by hopping through the native Function object (which will always give you the latest IL function). Multiple hops between IL functions are also safe: `low_level_il.medium_level_il.high_level_il`.

**Comment by wipawel** [comment]
I am actually working with the IL function object. I am not repeatedly grabbing it via native function. 
But, I do use `func_mlil.source_parameters` reference for the following situations:
1. I need to find cross references (call sites) for the `func_mlil`
2. I need to inspect list of input parameters to `func_mlil`
3. I need to get destination MLIL function object for a call inside `func_mlil`

It seems, this is where things become flaky?

Is there a way to do 1., 2. and 3. without going back to the native function reference?

**Comment by rssor** [comment]
Those sound like they should be safe. Going back to the native function object is fine, so long as you don’t use that native function object to go back to the IL functions (since they may not be the ones you came from) in a way that could result in mixing results from IL functions from different analysis updates of the same function.

The rest of this is written assuming that `func_mlil.source_parameters` there should be `func_mlil.source_function`.

By 3. do you mean get the MLIL function object for the call target of a call in some function? If so, that should be fine because the concern about mixing results between IL function objects is only relevant within a single function. Also another way to do that via the mlil call instruction directly might be `bv.get_function_at(mlil_call_inst.dest.constant)` (provided the call target is in fact constant).

So to go back to your original post, this bit stands out: `func.medium_level_il.ssa_form[x]` resulting in spurious failures. Because the IL function objects don’t change once they’re generated, I’m curious where `x` is coming from. Normally I’d expect that failing to be a result of getting an instruction index from one IL function object and using it in another accidentally (the most common form of this is accidentally mixing an index from an older IL function object, the underlying function being reanalyzed, then using the index with the most recent version). An index is only valid for the specific IL function object it came from since they’re not intended to be stable across function updates or forms. So if that’s not the problem we might be giving you an invalid instruction index from somewhere else and that’s what we need to chase down.

**Comment by wipawel** [comment]
I tracked the issue down to disappearing MLIL objects:
![image](https://user-images.githubusercontent.com/5213897/124572324-dce5ee80-de48-11eb-952a-e5828f390d5b.png)

The corresponding disassembly is there, the low level IL is there, but all of a sudden the MLIL is gone.

**Comment by rssor** [comment]
That's generally the result of the UI noticing that analysis is in progress and the object isn't ready yet, not that it's gone missing. We generate the ILs lazily on demand because they can be memory-intensive, so generally speaking when navigating to a function in the UI there's simply no IL to display at first.

Once you've got a IL function object that you're using nothing should happen to change it out from under you.

Do you by any chance have a script that can reproduce the behavior you're seeing when run on a sufficiently large binary?

**Comment by psifertex** [comment]
We've been unable to reproduce this and it doesn't sound like there are any other examples that can repro it at this point. 

For now we're going to close as "can't reproduce" but definitely feel free to re-open if you see it pop up again or get a good repro.
